### PR TITLE
Refactor decompress plugin to be a bit cleaner

### PR DESCRIPTION
**Pull request by ianstalk**
### Motivation for changes:
Clean up logic in the decompress plugin to make it cleaner and easier to follow. I've also been mulling over possibly making `decompress` an entry list plugin in the future, this should make that a bit easier.

### Detailed changes:
- Added ArchiveInfo class to `utils/archive.py`, which now does most of the heavy listing for decompressing files
- Moved additional logic into standalone functions

### Addressed issues:
- Fixes # .

### Implemented feature requests:
- Feathub #[XX](https://feathub.com/Flexget/Flexget/+XX).

### Config usage if relevant (new plugin or updated schema):
```
paste_config_here
```
### Log and/or tests output (preferably both):
```
flexget/tests/test_decompress.py .....                                                                                                                                                               [100%]

============================================================================================= warnings summary =============================================================================================
flexget/tests/test_decompress.py::TestExtract::()::test_rar
  /Users/ieaston/Flexget/flexget/tests/conftest.py:212: DeprecationWarning: use of getfuncargvalue is deprecated, use getfixturevalue
    dst = dst.replace('__tmp__', request.getfuncargvalue('tmpdir').strpath)
  /Users/ieaston/Flexget/flexget/tests/conftest.py:66: DeprecationWarning: use of getfuncargvalue is deprecated, use getfixturevalue
    config = config.replace('__tmp__', request.getfuncargvalue('tmpdir').strpath)

-- Docs: http://doc.pytest.org/en/latest/warnings.html
=================================================================================== 5 passed, 2 warnings in 2.54 seconds ===================================================================================
```

### CODE REVIEW

File: flexget/plugins/output/decompress.py [hunk 1]
```diff
@@ -25,53 +25,60 @@ def open_archive_entry(entry):
...
     """
-    arch = None
+    archive = None
 
     try:
         archive_path = entry['location']
-        arch = archive.open_archive(archive_path)
-    except KeyError:
+        if not archive_path:
+            raise ValueError()
```
Comment (liiight) [inline flexget/plugins/output/decompress.py]: Please add a message

File: flexget/plugins/output/decompress.py [hunk 2]
```diff
@@ -25,53 +25,60 @@ def open_archive_entry(entry):
...
-    """
-    Tests whether the file descibed in info is a directory
-    """
+    destination = get_destination_path(info, to, keep_dirs)
+
+    log.debug('Attempting to extract: %s to %s', info.filename, destination)
+    try:
+        info.extract(archive, destination)
+    except archiveutil.FSError as error:
+        log.error('OS error while creating file: %s (%s)' % (destination, error))
```
Comment (liiight) [inline flexget/plugins/output/decompress.py]: Should this/these raise PluginError instead?

File: flexget/plugins/output/decompress.py [hunk 3]
```diff
@@ -25,53 +25,60 @@ def open_archive_entry(entry):
...
-        return not base
 
+def get_destination_path(info, to, keep_dirs):
+    """Generate the destination path for a given file"""
 
-def makepath(path):
-    if not os.path.exists(path):
-        log.debug('Creating path: %s', path)
-        os.makedirs(path)
+    if keep_dirs:
```
Comment (liiight) [inline flexget/plugins/output/decompress.py]: style suggestion:
```python
path_suffix = info.path if keep_dirs else os.path.basename(info.path)
```

Comment (cvium) [reply flexget/plugins/output/decompress.py]: It's not. I find it more readable in this case though, but I don't care either way.

File: flexget/plugins/output/decompress.py (hunk 2 above)
Comment (ianstalk) [reply flexget/plugins/output/decompress.py]: Is PluginError fatal for the plugin? There's a chance that a FSError is related to a single file within an archive and the rest will extract successfully.

Comment (cvium) [reply flexget/plugins/output/decompress.py]: PluginError will abort the task

File: flexget/plugins/output/decompress.py (hunk 3 above)
Comment (ianstalk) [reply flexget/plugins/output/decompress.py]: That's two votes in favor of it, so I'll do it. :)

File: flexget/plugins/output/decompress.py [hunk 4]
```diff
@@ -28,13 +28,14 @@ def open_archive_entry(entry):
...
     try:
-        archive_path = entry['location']
-        if not archive_path:
-            raise ValueError('Entry does not appear to represent a local file.')
+        archive_path = entry.get('location')
 
-        archive = archiveutil.open_archive(archive_path)
-    except (KeyError, ValueError):
-        log.error('Entry does not appear to represent a local file.')
+        if not os.path.exists(archive_path):
```
Comment (cvium) [inline flexget/plugins/output/decompress.py]: This will raise an exception if archive_path is None

Comment (ianstalk) [reply flexget/plugins/output/decompress.py]: Addressed with my latest commit

Comment (cvium) [reply flexget/plugins/output/decompress.py]: Wouldn't it be better to reverse the if-else?
```
if not archive_path:
elif not os.path.exists
```

Comment (ianstalk) [reply flexget/plugins/output/decompress.py]: Not a bad idea. Let me change it.

**Comment by ianstalk** [comment]
@liiight Fixed.

**Pull request merged**